### PR TITLE
[Baseline 1] Add in/out estimation to Uniswap Pool struct

### PR DESCRIPTION
**Pull request by fleupold**
This PR implements the get_amounts_in/out methods based on the Uniswap smart contract logic. I added fee information to the pool struct, so that the in/out computation can be fee aware.

### Test Plan
Some unit tests bases on [this simulation](https://docs.google.com/spreadsheets/d/1K-tu5K43i3gBdlGs_LJNqgMGjEbm6j0Uf9YxxgoMJQg/edit#gid=0).

### CODE REVIEW

File: shared/src/uniswap_pool.rs [hunk 1]
```diff
@@ -19,10 +20,72 @@ pub trait PoolFetching: Send + Sync {
     async fn fetch(&self, token_pairs: HashSet<TokenPair>) -> Vec<Pool>;
 }
 
-#[derive(Clone)]
+#[derive(Clone, Hash)]
 pub struct Pool {
     pub tokens: TokenPair,
     pub reserves: (u128, u128),
+    pub fee: Rational,
```
Comment (vkgnosis) [inline shared/src/uniswap_pool.rs:27]: Rational is Ratio<isize>. The fee has to be positive so we could use Ratio<usize> or Ratio<u32>. Just pointing out, not super important.

File: shared/src/uniswap_pool.rs [hunk 2]
```diff
@@ -19,10 +20,72 @@ pub trait PoolFetching: Send + Sync {
...
+        } else {
+            assert!(token_in == self.tokens.get().1, "Token not part of pool");
+            (
+                U256::from(self.reserves.1),
+                U256::from(self.reserves.0),
+                self.tokens.get().0,
+            )
+        };
+
+        let amount_in_with_fee = amount_in * (self.fee.denom() - self.fee.numer());
```
Comment (vkgnosis) [inline shared/src/uniswap_pool.rs:57]: I think this and the other method are missing some overflow checks like in https://github.com/gnosis/gp-v2-services/blob/c7b301a4fdfe1f92b691a5e5bf64af4ecee31428/solver/src/uniswap.rs (form an unmerged PR) .
Might also be nice to have amount_out and amount_in from that file with the way the arguments don't include any H160. That way those methods can't panic and testing is a bit nicer and we have the match on the H160 separately from the math.

**Review by vkgnosis — APPROVED** [review]

File: shared/src/uniswap_pool.rs (hunk 2 above)
Comment (fleupold) [reply shared/src/uniswap_pool.rs:57]: 👍 on the missing overflow check. 

For the publicly facing interface, I find it nice to give callers the opportunity to get an amount without having to query the right reserve amount themselves from the outside. Would you suggest to split this into two methods then?

Comment (vkgnosis) [reply shared/src/uniswap_pool.rs:57]: Right you can still keep the current public method but have the inner in addition (maybe private).

**Pull request merged**